sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -619,3 +620,35 @@ object DataSourceStrategy {
     (nonconvertiblePredicates ++ unhandledPredicates, pushedFilters, handledFilters)
   }
 }
+
+
+/**
+ * Support for recalculating table statistics if table statistics are not available.
+ */
+class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
+
+  private val sessionConf = session.sessionState.conf
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    // For the data source table, we only recalculate the table statistics when it creates
+    // the CatalogFileIndex using defaultSizeInBytes. See SPARK-25474 for more details.
+    case logical @ LogicalRelation(_, _, Some(table), _)
+      if sessionConf.fallBackToHdfsForStatsEnabled && table.stats.isEmpty &&
+        sessionConf.manageFilesourcePartitions &&
+        table.tracksPartitionsInCatalog && table.partitionColumnNames.nonEmpty =>
+      val sizeInBytes = CommandUtils.getSizeInBytesFallBackToHdfs(session, table)
+      val withStats = table.copy(stats = Some(CatalogStatistics(sizeInBytes = BigInt(sizeInBytes))))
+      logical.copy(catalogTable = Some(withStats))
+
+    case relation: HiveTableRelation
```

Comment:
shall we catch `InsertIntoTable(HiveTableRelation)` as well?

Reply:
@advancedxy Already work on this:https://github.com/apache/spark/pull/25306/commits/c86a27b2a5e286733ad305de1d7e42d1373b3a3b